### PR TITLE
docs(contributing): add section on Design Documents

### PR DESCRIPTION
**Pull request by carmstrong**
Please review and criticize. I tried not to overuse the term "changes", but I wanted a word that was general enough to include any sort of change: features, major bug fixes, refactors, etc.

All input is welcome!

### CODE REVIEW

File: docs/contributing/design-documents.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
+:title: Design Documents
+:description: Information necessary for a feature design document.
+
+.. _design-documents:
+
+Design Documents
+================
+
+Before submitting a pull request which is expected to significantly change the
```
Comment (mboersma) [inline docs/contributing/design-documents.rst]: I would rephrase this because I'm a picky old newspaper copy editor:
> Before submitting a pull request which will change the behavior of Deis significantly, such as a new feature or major refactoring, contributors...

File: docs/contributing/design-documents.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+behavior of Deis (such as when contributing a new feature or refactor),
+contributors should first open an issue representing a design document.
+
+Goals
+-----
+
+Design documents were adopted in order to:
```
Comment (mboersma) [inline docs/contributing/design-documents.rst]: Maybe:
> Design documents help project contributors:

File: docs/contributing/design-documents.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+* Involve stakeholders as early as possible in a feature's development
+* Ensure code changes accomplish the original motivations and design goals
+* Establish clear acceptance criteria for a feature or change
+* Enforce test-driven design methodology and automated test coverage
```
Comment (mboersma) [inline docs/contributing/design-documents.rst]: Excellent point. The Deis project is going to stop making any exceptions for PRs without tests (which may require us to beef up our test coverage and create test harnesses in some places).

File: docs/contributing/design-documents.rst [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+Contents
+--------
+
+Design document issues should be named ``Design Doc: <change description>`` and
```
Comment (bacongobbler) [inline docs/contributing/design-documents.rst]: issue or pull request? If it's an issue, why not submit the design document in the original issue for a specific feature request? That way we have all design documents in one issue,  making it easier for external contributors to follow the conversation. OTOH, it makes the issue more convoluted, so I guess it's a question of whether github issues is the right tool for proposals.

File: docs/contributing/design-documents.rst [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+contain the following sections.
+
+Goal
+^^^^
+
+This section should briefly describe the proposed change and the motivations
+behind the change. Tests will be written to ensure this design goal is met by
```
Comment (mboersma) [inline docs/contributing/design-documents.rst]: "change" is repeated three times here, so a pronoun makes it read better:
> ... the motivations behind it.

File: docs/contributing/design-documents.rst [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+the change.
+
+This section will typically reference a separate GitHub issue tracking
+the feature or change, which will typically be assigned to a release milestone.
+
+Code Changes
+^^^^^^^^^^^^
+
+This section should detail the code changes necessary to accomplish the change,
```
Comment (bacongobbler) [inline docs/contributing/design-documents.rst]: +1. We could also make the commits themselves more verbose to detail the necessary changes.

File: docs/contributing/design-documents.rst [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+contain the following sections.
+
+Goal
+^^^^
+
+This section should briefly describe the proposed change and the motivations
+behind the change. Tests will be written to ensure this design goal is met by
+the change.
+
+This section will typically reference a separate GitHub issue tracking
```
Comment (mboersma) [inline docs/contributing/design-documents.rst]: I would omit the first "typically."

File: docs/contributing/design-documents.rst (hunk 6 above)
Comment (mboersma) [inline docs/contributing/design-documents.rst]: We could require that commits have a [message body](http://docs.deis.io/en/latest/contributing/standards/#message-body). I am often guilty of skipping that and just writing a subject line, but it's always useful to be more detailed.

File: docs/contributing/standards.rst [hunk 8]
```diff
@@ -82,18 +82,6 @@ the exception that lines of code may be up to 99 characters long. Docstrings and
 tests are required for all public methods, although the flake8_ tool used by
 Deis does not enforce this.
 
-
-Design Proposals
```
Comment (bacongobbler) [inline docs/contributing/standards.rst]: I'd add a link back in this doc to the new design proposal document, as you did with the contributing overview.

File: docs/contributing/design-documents.rst [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+^^^^^
+
+All changes should be covered by automated tests, either unit or integration tests
+(ideally both). This section should detail how tests will be written to validate
+that the change accomplishes the design goals and doesn't introduce any regressions.
+
+If a change cannot be sufficiently covered by automated testing, the design
+should be reconsidered. If there is no test coverage whatsoever for an affected
+section of code, a separate issue should be filed to integrate automated testing
+with that section of the codebase.
```
Comment (mboersma) [inline docs/contributing/design-documents.rst]: Yes--if it's unreasonable to write a test for the feature because there isn't any example already of how to do so, that's a high-priority issue for Deis now that we must fix now.

**Comment by bacongobbler** [comment]
So with this PR, we're changing the terminology from a "Proposal" to a "Design Document". I like it. A "proposal" was loosely termed, so we found that most proposals were pretty sparse on the details of how a feature would be implemented, the motivation behind changes etc. :+1:

File: docs/contributing/design-documents.rst (hunk 3 above)
Comment (bacongobbler) [inline docs/contributing/design-documents.rst]: +1

File: docs/contributing/design-documents.rst (hunk 4 above)
Comment (carmstrong) [inline docs/contributing/design-documents.rst]: I see the general tracking issue being a typically short, high-level description of an issue - such as "we need automated tests for the client". This issue would be assigned a milestone, but wouldn't really be fleshed out in much detail. When someone begins work, they'd open a design document to flesh out exactly how they're going about the implementation of the issue.

I can't really formulate a strong argument as to why it should be a separate issue, other than that it *feels* like it'd get more attention and review. But there are benefits to keeping it in the same issue, such as watchers of that issue being notified when there's an update. I'm open to discussion on this.

File: docs/contributing/design-documents.rst (hunk 6 above)
Comment (carmstrong) [inline docs/contributing/design-documents.rst]: Commit bodies are good, but this is meant to be more of a discussion, forcing people to think through the change - "we need to change this code because of how it's used here, meaning it should be refactored as an interface".

**Comment by carmstrong** [comment]
> So with this PR, we're changing the terminology from a "Proposal" to a "Design Document". I like it.

Yes and no - I think proposals could also be considered "tracking issues". Copying a comment I made inline here:

I see the general tracking issue being a typically short, high-level description of an issue - such as "we need automated tests for the client". This issue would be assigned a milestone, but wouldn't really be fleshed out in much detail. When someone begins work, they'd open a design document to flesh out exactly how they're going about the implementation of the issue.

I can't really formulate a strong argument as to why it should be a separate issue, other than that it feels like it'd get more attention and review. But there are benefits to keeping it in the same issue, such as watchers of that issue being notified when there's an update. I'm open to discussion on this.

File: docs/contributing/design-documents.rst (hunk 6 above)
Comment (mboersma) [inline docs/contributing/design-documents.rst]: Agreed--I guess I was suggesting that maintainers could agree to insist on more complete commit messages, but that's off-topic for this PR.

**Comment by rogerleite** [comment]
You could mention this document on CONTRIBUTING.md

File: docs/contributing/design-documents.rst (hunk 4 above)
Comment (bacongobbler) [inline docs/contributing/design-documents.rst]: I think it's fine to keep them as separate issues for now, though we can experiment with in-line design docs in the future if we're failing to get feedback. :)

File: docs/contributing/design-documents.rst (hunk 6 above)
Comment (carmstrong) [inline docs/contributing/design-documents.rst]: > I guess I was suggesting that maintainers could agree to insist on more complete commit messages

:100: 



> but that's off-topic for this PR.

It is, but I'd suggest opening a separate issue to track that. I think it's good practice.

File: docs/contributing/standards.rst (hunk 8 above)
Comment (carmstrong) [inline docs/contributing/standards.rst]: I did earlier.

**Comment by carmstrong** [comment]
> You could mention this document on CONTRIBUTING.md

Good idea - I'll do that.

**Comment by carmstrong** [comment]
Updated CONTRIBUTING.md. Thanks @rogerleite!

**Comment by mboersma** [comment]
LGTM.

**Comment by bacongobbler** [comment]
docs LGTM
